### DISCUSSION TITLE
Difference between the k8s Gracefull Node Shutdow

### DISCUSSION THREAD

**Discussion by tete17** [aws/aws-node-termination-handler#457] [Q&A] [answered]
I recently was confronted with the unfortunate scenario of one of my EC2 VM to be marked for a maintenance window and needing to be restarted. Lazy email reader as I am I forgot and my VM got restarted for me causing a small but noticeable downtime until the control plane detected the node being down. 

I recently found this cool project to alleviate part of my issues since I run in AWs but I also found that very recently k8s has launched this new feature https://kubernetes.io/blog/2021/04/21/graceful-node-shutdown-beta/ that seems to target the same goals as this project, graceful shutdown and spot instance support. I am trying to decide if it is still worth it to install this helm chart into my cluster after upgrading to k8s 1.21 and get this feature for free. I am guessing this project has some other advantages (other than backporting this feature to older k8s clusters) I am not able to destil from its README. Could you guys enlighten me please?

**Comment by bwagner5** [comment, marked as answer]
Hello! Yes, the aws-node-termination-handler does offer additional features on-top of what the k8s graceful node shutdown will give you. One such benefit is the time you have to drain your spot instance. The k8s graceful node shutdown will not give you the full 2 mins that a spot interruption termination notice will give you (which the aws-node-termination-handler monitors and handles). NTH also supports Spot Rebalance Recommendation which can give you even more time to drain a node and will also allow the ASG to backfill your instance before it is terminated resulting in a much faster pod migration (you could enable rebalance w/ k8s grace node shutdown as well, you would just lose the lifecycle hook completions which would limit the time your instance would have to shutdown). NTH also offers webhook notifications for interruption events.

If you do not need much time to drain your nodes and don't care to get notified, then the k8s graceful node shutdown may be good enough for your use-case.
